src/bsmHandler.cpp

```diff
@@ -222,58 +226,76 @@ bool BSMHandler::process( const std::string& message_json ) {
 
         rapidjson::Value& data = payload["data"];
 
-        if (!data.HasMember("coreData")) {
+        if (!data.HasMember("value")) {
+            result_ = ResultStatus::MISSING;
+
+            return false;
+        }
+
+        rapidjson::Value& value = data["value"];
+
+        if (!value.HasMember("BasicSafetyMessage")) {
             result_ = ResultStatus::MISSING;
 
             return false;
         }
 
-        rapidjson::Value& core_data = data["coreData"];
+        rapidjson::Value& basicSafetyMessage = value["BasicSafetyMessage"];
+
+        if (!basicSafetyMessage.HasMember("coreData")) {
+            result_ = ResultStatus::MISSING;
+
+            return false;
+        }
+
+        rapidjson::Value& core_data = basicSafetyMessage["coreData"];
 
         if (!core_data.HasMember("speed")) {
             result_ = ResultStatus::MISSING;
 
             return false;
         }
         
-        if (!core_data["speed"].IsDouble()) {
+        if (!core_data["speed"].IsInt()) {
             result_ = ResultStatus::OTHER;
 
             return false;
         }
 
-        speed = core_data["speed"].GetDouble();
-        bsm_.set_velocity(speed);
+        // Only set speed if it's not the J2735 "unavailable" value
+        if (core_data["speed"].GetInt() != J2735_SPEED_UNAVAILABLE) {
+            speed = core_data["speed"].GetInt() * 0.02;
```

Comment:
suggestion: add comment explaining why we're multiplying by the .02 value

Reply:
This technically has the comment that explains the value used as well.

Reply:
Addressed.